### PR TITLE
encrypt and decrypt tokens

### PR DESCRIPTION
**Pull request by mitchellhenke**
Happy path token encrypting/decrypting

### CODE REVIEW

File: lib/identity_pki/token.ex [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  def open(token, key \\ secret_key()) do
+    # TODO: check expires_at
+    # TODO: check HMAC
+    [text, iv, tag] =
+      String.split(token, "--")
+      |> IO.inspect(label: "SPLIT TOKEN")
+      |> Enum.map(&Base.decode64!/1)
+
+    :crypto.crypto_one_time_aead(:aes_256_gcm, key, iv, text, "", tag, false)
+    |> Jason.decode!()
```
Comment (zachmargolis) [inline lib/identity_pki/token.ex:29]: omg they called the JSON library Jason 😂 😭

Comment (mitchellhenke) [reply lib/identity_pki/token.ex:29]: It's great!

File: lib/identity_pki/token.ex [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    data =
+      if expires_at do
+        Map.put(data, :expires_at, expires_at)
+      else
+        data
+      end
+
+    json =
+      Map.put(data, :r1, random_bytes(8))
```
Comment (mitchellhenke) [inline lib/identity_pki/token.ex:14]: Ruby PKI takes advantage of Ruby's hash keys being encoded in order and makes sure that `r1` is the first key and `r2` is the last key to surround the rest of the data as entropy. Elixir maps don't have a deterministic key order so to more closely match that we could do `Jason.encode!([r1, data, r2])`, but we weren't sure if that was sufficient.

File: lib/identity_pki/token.ex [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      :crypto.crypto_one_time_aead(:aes_256_gcm, key, iv, data, "", 16, true)
+
+    [Base.encode64(cipher_text), Base.encode64(iv), Base.encode64(cipher_tag)]
+    |> Enum.join("--")
+  end
+
+  defp secret_key do
+    salt = Application.get_env(:identity_pki, :token_encryption_key_salt)
+    pepper = Application.get_env(:identity_pki, :token_encryption_key_pepper)
+    Plug.Crypto.KeyGenerator.generate(pepper, salt, digest: :sha, iterations: 65_536)
```
Comment (mitchellhenke) [inline lib/identity_pki/token.ex:46]: Rails PKI uses `ActiveSupport::KeyGenerator`, which defaults to the above options. Confirmed that the same pepper and salt result in the same key being generated in both. This is currently re-generating the key every time, which is slow (Rails caches the result).

File: lib/identity_pki_web/controllers/page_controller.ex [hunk 4]
```diff
@@ -14,36 +13,27 @@ defmodule IdentityPkiWeb.PageController do
...
-      uuid: "c229c00a-5b80-424c-aa5c-f5a0e97e9fa1",
-      card_type: "piv",
-      auth_cert: Certificate.auth_cert?(cert),
-      nonce: nonce
-    }
+    token =
+      %{
+        subject: Certificate.rfc_2253_subject(cert),
+        issuer: Certificate.issuer(cert),
+        uuid: "c229c00a-5b80-424c-aa5c-f5a0e97e9fa1",
```
Comment (mitchellhenke) [inline lib/identity_pki_web/controllers/page_controller.ex:20]: This should be unique for each PIV/CAC

**Review by orenyk — APPROVED** [review]
:shipit:  🦉

**Pull request merged**